File: 5Days-SoftwareArchitecture/typesript/Day1- Metrics/ReflectionDemo.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var Student = /** @class */ (function () {
    function Student() {
    }
    Student.prototype.study = function () {
        console.log("Student studies general subjects");
    };
    return Student;
}());
var ScienceStudent = /** @class */ (function (_super) {
    __extends(ScienceStudent, _super);
    function ScienceStudent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ScienceStudent.prototype.study = function () {
        console.log("Science student studies Physics and Chemistry");
    };
    return ScienceStudent;
}(Student));
var ArtsStudent = /** @class */ (function (_super) {
    __extends(ArtsStudent, _super);
    function ArtsStudent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ArtsStudent.prototype.study = function () {
        console.log("Arts student studies History and Literature");
    };
    return ArtsStudent;
}(Student));
var s1 = new ScienceStudent();
var s2 = new ArtsStudent();
console.log("--- Normal Virtual Call ---");
var s = s1;
s.study();
console.log("\n--- Reflection Call ---");
// Simulate runtime lookup using dynamic method call
s = Math.random() > 0.5 ? s1 : s2;
s["study"](); // TypeScript dynamic call, like reflection

File: 5Days-SoftwareArchitecture/Java/Day1 CC Metrics/ReflectionDemo.java
import java.lang.reflect.Method;

// Base class
class Student {
    public void study() {
        System.out.println("Student studies general subjects");
    }
}

// Science student
class ScienceStudent extends Student {
    @Override
    public void study() {
        System.out.println("Science student studies Physics and Chemistry");
    }
}

// Arts student
class ArtsStudent extends Student {
    @Override
    public void study() {
        System.out.println("Arts student studies History and Literature");
    }
}

public class ReflectionDemo {
    public static void main(String[] args) throws Exception {
        Student s1 = new ScienceStudent();
        Student s2 = new ArtsStudent();

        System.out.println("--- Normal Virtual Call ---");
        // Always use ScienceStudent for normal call
        Student s = s1;
        s.study();  // virtual method call, may be optimized by JVM

        System.out.println("\n--- Reflection Call ---");
        // Reflection forces runtime lookup, still keep randomness
        s = Math.random() > 0.5 ? s1 : s2;
        Method m = s.getClass().getMethod("study"); // find method at runtime
        m.invoke(s);  // runtime method lookup, cannot optimize
    }
}
